### PR TITLE
Fix #153/#154 : MultipleChoice/ItemSelection Interaction 

### LINKED ISSUE #153: UI Structure: MultipleChoiceInteractionView [Blocked: #150]
This is tracking implementing the initial UI structure of the multiple choice interaction being fully implemented in #31.

### PR DESCRIPTION
**Pull request by veena14cs**
## Explanation
This PR is based on content-card #275  , as it contains Html parser implementation and it is used in MultipleChoice/ItemSelection Input interaction.

This PR is duplicate of #258 . It contains UI structure for Multiple choice and item selection interactions, and fixes #153 and #154.

Reference Design Doc: https://docs.google.com/document/d/1HFwDFcqenAXepbYr5rfawGHvkeBZpYvNMs228AWLn24/edit?usp=sharing

### CODE REVIEW

File: app/src/main/java/org/oppia/app/player/state/InteractionAdapter.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/player/state/InteractionAdapter.kt]: Nit, the formatting here should be:

```kotlin
} else {
```

Ditto elsewhere below.

Please use the code reformat option in Android Studio to fix this & other formatting discrepancies.

**Comment by BenHenning** [comment]
Please reassign me once the base PR is merged in.

**Comment by rt4914** [comment]
@veena14cs Assign it back to me once the approach in this PR is similar to that of #275

**Comment by veena14cs** [comment]
@BenHenning I am facing race condition error, also it was mentioned by @rt4914. Can please take a look into it, as I am unable to use exploration.

Screenshot 2019-11-04 at 5 26 32 PM

File: app/src/main/java/org/oppia/app/player/state/InteractionAdapter.kt
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/state/InteractionAdapter.kt]: It should be `*/`

File: app/src/main/java/org/oppia/app/player/state/InteractionAdapter.kt
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/state/InteractionAdapter.kt:33]: `Optional Nit`: Press Option+Cmd + L to fix some nit changes in this file

File: app/src/main/java/org/oppia/app/player/state/InteractionAdapter.kt
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/state/InteractionAdapter.kt]: Update this with `else -> throw IllegalArgumentException("Invalid view type: $viewType")` for better error

File: app/src/main/java/org/oppia/app/player/state/InteractionAdapter.kt
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/state/InteractionAdapter.kt]: Replace this with logger.

File: app/src/main/java/org/oppia/app/player/state/InteractionAdapter.kt
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/state/InteractionAdapter.kt]: Remove this blank line.

File: app/src/main/java/org/oppia/app/player/state/SelectInputItemsListener.kt
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/state/SelectInputItemsListener.kt:5]: Incorrect comment.

File: app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt:22]: Nit: Option + Command + L

**Review by rt4914 — CHANGES_REQUESTED** [review]
@veena14cs Please mention the steps to test this PR because I tried launching ContentCardTestActivity but it is not working.

File: app/src/main/java/org/oppia/app/player/state/InteractionAdapter.kt
Comment (rt4914) [reply app/src/main/java/org/oppia/app/player/state/InteractionAdapter.kt]: Done.

File: app/src/main/java/org/oppia/app/player/state/InteractionAdapter.kt
Comment (rt4914) [reply app/src/main/java/org/oppia/app/player/state/InteractionAdapter.kt:33]: Done

File: app/src/main/java/org/oppia/app/player/state/InteractionAdapter.kt
Comment (rt4914) [reply app/src/main/java/org/oppia/app/player/state/InteractionAdapter.kt]: Done.

File: app/src/main/java/org/oppia/app/player/state/InteractionAdapter.kt
Comment (rt4914) [reply app/src/main/java/org/oppia/app/player/state/InteractionAdapter.kt]: Done.

File: app/src/main/java/org/oppia/app/player/state/InteractionAdapter.kt
Comment (rt4914) [reply app/src/main/java/org/oppia/app/player/state/InteractionAdapter.kt]: Done.

File: app/src/main/java/org/oppia/app/player/state/SelectInputItemsListener.kt
Comment (rt4914) [reply app/src/main/java/org/oppia/app/player/state/SelectInputItemsListener.kt:5]: Changed comment.

**Comment by rt4914** [comment]
> @veena14cs Please mention the steps to test this PR because I tried launching ContentCardTestActivity but it is not working.

Solved this issues and tested the app.

**Comment by rt4914** [comment]
@BenHenning I have made changes to this PR to make it ready for you.

This PR uses same approach as #275 and also there is this same issue of race-condition in `ExplorationDataController`.
Other than that all the test cases are passing.

**Comment by rt4914** [comment]
@BenHenning merge conflicts have been resolved.

File: app/build.gradle
Comment (BenHenning) [inline app/build.gradle:42]: Awesome! I was wondering if this was possible via Gradle. Great!

File: app/src/main/java/org/oppia/app/player/state/listener/ItemClickListener.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/player/state/listener/ItemClickListener.kt]: Nit: space before ``{``.

File: app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentTest.kt
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentTest.kt]: What purpose do these orientation changes serve? If we want to test a single orientation change, we should just do one rather than 2 here.

File: app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentTest.kt
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentTest.kt]: Does this actually check the number of selections? It seems like the above could be radio buttons and this would still be counting for each interaction that's successful. I suggest removing this, and if we care to check the number of selections we should probably rely on the submitted answer.

File: build.gradle
Comment (BenHenning) [inline build.gradle:8]: What is this for?

File: domain/src/main/java/org/oppia/domain/exploration/ExplorationRetriever.kt
Comment (BenHenning) [inline domain/src/main/java/org/oppia/domain/exploration/ExplorationRetriever.kt:26]: Please introduce a new ID rather than overwriting this--other tests depend on this exploration being about_oppia.

Also, please make sure you're running the domain tests before submitting PRs--I didn't catch this until I started working on tests for #270.

File: domain/src/main/java/org/oppia/domain/util/StateRetriever.kt
Comment (BenHenning) [inline domain/src/main/java/org/oppia/domain/util/StateRetriever.kt]: This isn't correct--ItemSelection uses type StringList, not negative int. Please fix.

**Review by BenHenning — APPROVED** [review]
Thanks @veena14cs. I glanced through all of it, but didn't evaluate the design in detail since much of it is changing in #270 to address issues I noticed in my original recommended design for the exploration player. I did notice a few key issues we should address before merging.

File: build.gradle
Comment (veena14cs) [reply build.gradle:8]: To disable animation in Espresso testing need to add plugin.
https://stackoverflow.com/questions/43751079/espresso-testing-disable-animation/43861179#43861179.

File: app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentTest.kt
Comment (veena14cs) [reply app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentTest.kt]: okay it was added while debugging. Removed it.

File: app/src/main/java/org/oppia/app/player/state/listener/ItemClickListener.kt
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/state/listener/ItemClickListener.kt]: Done.

File: app/src/main/java/org/oppia/app/player/state/InteractionAdapter.kt
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/state/InteractionAdapter.kt]: Done.

File: build.gradle
Comment (BenHenning) [reply build.gradle:8]: Ah, can you add a brief comment in the Gradle file providing context & a link to that Stackoverflow answer? I think that would be helpful to other folks who run across this code and are unsure of what it's for.

**Comment by nikitamarysolomanpvt** [comment]
Please do re assign when the changes are done

File: build.gradle
Comment (veena14cs) [reply build.gradle:8]: Yes sure I will add comment thanks.

File: domain/src/main/java/org/oppia/domain/util/StateRetriever.kt
Comment (veena14cs) [reply domain/src/main/java/org/oppia/domain/util/StateRetriever.kt]: @BenHenning  please check this implementation.

File: domain/src/main/java/org/oppia/domain/util/StateRetriever.kt
Comment (BenHenning) [inline domain/src/main/java/org/oppia/domain/util/StateRetriever.kt]: This doesn't seem quite right to me--the key should correspond to an array not an int. See: https://github.com/oppia/oppia-android/pull/270/files#diff-8ee6f925fbb2de9ecb09094ccf9ac5cdR169.

File: domain/src/main/java/org/oppia/domain/exploration/ExplorationRetriever.kt
Comment (veena14cs) [reply domain/src/main/java/org/oppia/domain/exploration/ExplorationRetriever.kt:26]: reverted TEST_EXPLORATION_ID_6 to `about_oppia.json`

Comment (veena14cs) [reply domain/src/main/java/org/oppia/domain/exploration/ExplorationRetriever.kt:26]: Added new Exploration id for ItemSelectionInput.

File: domain/src/main/java/org/oppia/domain/util/StateRetriever.kt
Comment (veena14cs) [reply domain/src/main/java/org/oppia/domain/util/StateRetriever.kt]: Fixed this thanks.

File: domain/src/main/java/org/oppia/domain/util/StateRetriever.kt
Comment (veena14cs) [reply domain/src/main/java/org/oppia/domain/util/StateRetriever.kt]: Changed implementation according to #270.

File: app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentTest.kt
Comment (veena14cs) [reply app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentTest.kt]: Removed this assert check.

File: app/build.gradle
Comment (veena14cs) [reply app/build.gradle:42]: Thanks but need to test with other devices too.

**Comment by BenHenning** [comment]
Thanks @veena14cs! I looked at the fixes--they seem sensible and compatible with #270. Please go ahead and submit this pending @rt4914's approval.

**Comment by veena14cs** [comment]
> Thanks @veena14cs! I looked at the fixes--they seem sensible and compatible with #270. Please go ahead and submit this pending @rt4914's approval.

Okay, I will merge once @rt4914  approves it. Thanks.

**Review by rt4914 — APPROVED** [review]
LGTM, thanks @veena14cs and @BenHenning

**Pull request merged**
